File: libs/sdk/android/src/main/java/com/rudderstack/react/android/RNParamsConfigurator.java
package com.rudderstack.react.android;

import com.facebook.react.bridge.ReadableMap;
import com.rudderstack.android.sdk.core.RudderConfig;

class RNParamsConfigurator {
    private final ReadableMap config;
    boolean trackLifeCycleEvents = true;
    boolean recordScreenViews = false;
    long sessionTimeout = 300000L;
    boolean autoSessionTracking = true;
    String writeKey;

    RNParamsConfigurator(ReadableMap config) {
        this.config = config;
    }

    RudderConfig.Builder handleConfig() {
        setConfigValues();
        setWriteKey();
        RudderConfig.Builder configBuilder = buildConfig();
        disableAutoConfigFlagsForNativeSDK(configBuilder);
        return configBuilder;
    }

    private void setWriteKey() {
        if (this.config.hasKey("writeKey")) {
            writeKey = config.getString("writeKey");
        } else {
            throw new IllegalArgumentException("writeKey is required");
        }
    }

    private void setConfigValues() {
        if (this.config.hasKey("trackAppLifecycleEvents")) {
            trackLifeCycleEvents = config.getBoolean("trackAppLifecycleEvents");
        }
        if (this.config.hasKey("recordScreenViews")) {
            recordScreenViews = config.getBoolean("recordScreenViews");
        }
        if (this.config.hasKey("sessionTimeout")) {
            sessionTimeout = (long) config.getDouble("sessionTimeout");
        }
        if (this.config.hasKey("autoSessionTracking")) {
            autoSessionTracking = config.getBoolean("autoSessionTracking");
        }
    }

    private RudderConfig.Builder buildConfig() {
        RudderConfig.Builder configBuilder = new RudderConfig.Builder();
        if (config.hasKey("dataPlaneUrl")) {
            configBuilder.withDataPlaneUrl(config.getString("dataPlaneUrl"));
        }
        if (config.hasKey("controlPlaneUrl")) {
            configBuilder.withControlPlaneUrl(config.getString("controlPlaneUrl"));
        }
        if (config.hasKey("flushQueueSize")) {
            configBuilder.withFlushQueueSize(config.getInt("flushQueueSize"));
        }
        if (config.hasKey("dbCountThreshold")) {
            configBuilder.withDbThresholdCount(config.getInt("dbCountThreshold"));
        }
        if (config.hasKey("sleepTimeOut")) {
            configBuilder.withSleepCount(config.getInt("sleepTimeOut"));
        }
        if (config.hasKey("configRefreshInterval")) {
            configBuilder.withConfigRefreshInterval(config.getInt("configRefreshInterval"));
        }
        if (config.hasKey("autoCollectAdvertId")) {
            configBuilder.withAutoCollectAdvertId(config.getBoolean("autoCollectAdvertId"));
        }
        if (config.hasKey("logLevel")) {
            configBuilder.withLogLevel(config.getInt("logLevel"));
        }
        if (config.hasKey("collectDeviceId")) {
            configBuilder.withCollectDeviceId(config.getBoolean("collectDeviceId"));
        }
        return configBuilder;
    }

    private void disableAutoConfigFlagsForNativeSDK(RudderConfig.Builder configBuilder) {
        // We are relying on the RN implementation in RNLifeCycleEventListener and RNUserSessionPlugin hence we are explicitly setting all flags to false in Native Android SDK
        configBuilder.withRecordScreenViews(false);
        configBuilder.withTrackLifecycleEvents(false);
        configBuilder.withAutoSessionTracking(false);
    }
}

File: libs/sdk/ios/RNParamsConfigurator.m
//
//  RNParamsConfigurator.m
//  RNRudderSdk
//
//  Created by Abhishek Pandey on 02/08/23.
//

#import <Foundation/Foundation.h>
#import "RNParamsConfigurator.h"
#import "RNRudderAnalytics.h"

@implementation RNParamsConfigurator

- (instancetype) initWithConfig:(NSDictionary*)config {
    self = [super init];
    if (self) {
        self->config = config;

        // Default values
        self.trackLifeCycleEvents = true;
        self.recordScreenViews = false;
        self.sessionTimeout = 300000L;
        self.autoSessionTracking = true;
        self.enableBackgroundMode = false;
        self->dbEncryption = nil;
    }
    return self;
}

-(RSConfigBuilder*)handleConfig {
    [self setConfigValues];
    [self setWriteKey];
    RSConfigBuilder *configBuilder = [self buildConfig];
    [self addDBEncryptionPluginIfAvailable:configBuilder];
    [self disableAutoConfigFlagsForNativeSDK:configBuilder];
    return configBuilder;
}

-(void)setConfigValues {
    if ([config objectForKey:@"trackAppLifecycleEvents"]) {
        self.trackLifeCycleEvents =  [config[@"trackAppLifecycleEvents"] boolValue];
    }
    if ([config objectForKey:@"recordScreenViews"]) {
        self.recordScreenViews = [config[@"recordScreenViews"] boolValue];
    }
    if ([config objectForKey:@"sessionTimeout"]) {
        self.sessionTimeout = [config[@"sessionTimeout"] intValue];
    }
    if ([config objectForKey:@"autoSessionTracking"]) {
        self.autoSessionTracking = [config[@"autoSessionTracking"] boolValue];
    }
    if ([config objectForKey:@"enableBackgroundMode"]) {
        self.enableBackgroundMode = [config[@"enableBackgroundMode"] boolValue];
    }
}

-(void)setWriteKey {
    self.writeKey = config[@"writeKey"];
}

-(RSConfigBuilder*)buildConfig {
    RSConfigBuilder* configBuilder = [[RSConfigBuilder alloc] init];
    if ([config objectForKey:@"dataPlaneUrl"]) {
        [configBuilder withDataPlaneUrl:config[@"dataPlaneUrl"]];
    }
    if ([config objectForKey:@"controlPlaneUrl"]) {
        [configBuilder withControlPlaneUrl:config[@"controlPlaneUrl"]];
    }
    if ([config objectForKey:@"flushQueueSize"]) {
        [configBuilder withFlushQueueSize:[config[@"flushQueueSize"] intValue]];
    }
    if ([config objectForKey:@"dbCountThreshold"]) {
        [configBuilder withDBCountThreshold:[config[@"dbCountThreshold"] intValue]];
    }
    if ([config objectForKey:@"sleepTimeOut"]) {
        [configBuilder withSleepTimeOut:[config[@"sleepTimeOut"] intValue]];
    }
    if ([config objectForKey:@"configRefreshInterval"]) {
        [configBuilder withConfigRefreshInteval:[config[@"configRefreshInterval"] intValue]];
    }
    if ([config objectForKey:@"logLevel"]) {
        [configBuilder withLoglevel:[config[@"logLevel"] intValue]];
    }
    if ([config objectForKey:@"collectDeviceId"]) {
        [configBuilder withCollectDeviceId:[config[@"collectDeviceId"] boolValue]];
    }
    if ([config objectForKey:@"enableGzip"]) {
      [configBuilder withGzip:[config[@"enableGzip"] boolValue]];
    }
    return configBuilder;
}

-(void)addDBEncryptionPluginIfAvailable:(RSConfigBuilder *)configBuilder {
    self->dbEncryption = [RNRudderAnalytics getDBEncryption];
    if (self->dbEncryption) {
        [configBuilder withDBEncryption:self->dbEncryption];
    }
}

-(void) disableAutoConfigFlagsForNativeSDK:(RSConfigBuilder*)configBuilder {
    [configBuilder withTrackLifecycleEvens:false];
    [configBuilder withRecordScreenViews:false];
    [configBuilder withAutoSessionTracking:false];
    [configBuilder withEnableBackgroundMode:false];
}

@end

